qiskit_experiments/calibration_management/backend_calibrations.py

```diff
@@ -196,7 +213,149 @@ def export_backend(self) -> Backend:
 
         backend.defaults().qubit_freq_est = self.get_qubit_frequencies()
         backend.defaults().meas_freq_est = self.get_meas_frequencies()
-
-        # TODO: build the instruction schedule map using the stored calibrations
+        backend.default().instruction_schedule_map = self._inst_map
 
         return backend
+
+    def inst_map_add(
+        self,
+        instruction_name: str,
+        qubits: Tuple[int],
+        schedule_name: Optional[str] = None,
+        assign_params: Optional[Dict[Union[str, ParameterKey], ParameterValueType]] = None,
+    ):
+        """Update a single instruction in the instruction schedule map.
+
+        This method can be used to update a single instruction for the given qubits but
+        it can also be used by experiments that define custom gates with parameters
+        such as the :class:`Rabi` experiment. In a Rabi experiment there is a gate named
+        "Rabi" that scans a pulse with a custom amplitude. Therefore we would do
+
+        .. code-block:: python
+
+            cals.inst_map_add("Rabi", (0, ), "xp", assign_params={"amp": Parameter("amp")})
+
+        to temporarily add a pulse for the Rabi gate in the instruction schedule map. This
+        then allows calling :code:`transpile(circ, inst_map=cals.instruction_schedule_map)`.
+
+        Args:
+            instruction_name: The name of the instruction to add to the instruction schedule map.
+            qubits: The qubits to which the instruction will apply.
+            schedule_name: The name of the schedule. If None is given then we assume that the
+                schedule and the instruction have the same name.
+            assign_params: An optional dict of parameter mappings to apply. See for instance
+                :meth:`get_schedule` of :class:`Calibrations`.
+        """
+        schedule_name = schedule_name or instruction_name
+
+        inst_map_args = None
+        if assign_params is not None:
+            inst_map_args = assign_params.keys()
+
+        self._inst_map.add(
+            instruction=instruction_name,
+            qubits=qubits,
+            schedule=self.get_schedule(schedule_name, qubits, assign_params),
+            arguments=inst_map_args,
+        )
+
+    def inst_map_remove(self, instruction: str, qubits: Tuple[int]):
+        """Remove a single instruction from the inst_map."""
+        self._inst_map.remove(instruction, qubits)
+
+    def complete_inst_map_update(self, schedules: Optional[set] = None):
```

Comment:
We could forget calling this method and may run into a silent bug (i.e. calibrated parameter doesn't appear in other circuits). What is the motivation of separating this from the instmap getter method?

Reply:
The rational here is efficiency. Updating the whole `inst_map` each time we call `cals.instruction_schedule_map` is inefficient. E.g. if I have many qubits and run a cal on say qubit 0. Then I don't want to have to update the whole `inst_map` for all qubits when I want to get the `inst_map` for the next cal experiment. My thinking is the following:
* Upon `__init__` of the `cals` update the whole inst map.
* Each time a parameter value is updated in the `cals` then update only the affected schedules in the `inst_map`.

The gain here is efficiency, the tradeoff is that we need to be diligent with the `inst_map` update. This is why `BackendCalibrtions` now enforces an `inst_map` update on the concerned schedules. This makes me think that I should rename `complete_inst_map_update` to `inst_map_update` since it can specify which schedule to update. Also specifying the qubits to update might make sense.

Reply:
This commit 956a932 improved the inst. map updating and added a test for the expected behavior.

Reply:
That make sense. Thanks Daniel. What do you think tuning `_inst_map_update` into a protected method? Is there any situation user wants to call it explicitly?

Reply:
That sounds tempting but after some reflection I think there is a rational to keep it public. Consider the situation where the user adds his own schedule to the cals, i.e. he does not use the library and the init method. The workflow is the following:
1. Add the schedule to the cals with `add_schedule`. At this point we cannot yet add it to the inst map as the schedule may well be fully parameterized.
2. Add values for the parameters (either qubit specific or default values).
3. Update the inst. map by calling `update_inst_map({new_sched_name})`. This step can only be done after there are values in the parameters otherwise the schedule will not be added.

Based on this case I would leave `update_inst_map` public for now.